### ISSUE TITLE
feature request: test if passing expression to empty()

### ISSUE THREAD

**Issue by bkdotcom** [phpmd/phpmd#431]
http://us3.php.net/manual/en/function.empty.php
> Prior to PHP 5.5, empty() only supports variables; anything else will result in a parse error. In other words, the following will not work: `empty(trim($name))`. Instead, use `trim($name) == false`.

would it be possible to test if an expression is being passed to empty()?

but still allow vars/properties such as `empty(Yii::app()->user->id)` and `empty(self::$myProperty)`

unit-testing *should* catch `empty(getSomeValue())` from going to prod in a 5.4 environment... we all have unit-testing right?  ☺

**Comment by ravage84** [comment]
Hi @bkdotcom 

Unfortunately, we cannot test on that as PHPMD is a static code analysis tool. We do not execute the code and getting the information out of the code examples you provided, would depend on executing the code.

**Comment by bkdotcom** [comment]
Don't mistake my request as a request to catch runtime errors like
```php
// I don't expect/want phpmd to catch this
$var = null;
$var->foo; // Notice: Trying to get property of non-object
```

Nor am I expecting phpmd to catch undefined property/method type runtime errors.

If a visual code inspection _(static analysis performed by human)_ can determine validity then static analysis performed by algorithm should be able to as well 

I don't know what type of static analysis tools are at our disposal.
This is likely a "is it worth the effort/expense" question

This isn't my area of expertise at all, but I cobbled together a big ugly regex that seems to do a decent job for a first effort.

```php

/**
 * nothing at runtime (that I can think of) will change whether or not these are "valid" / "invalid"
 *  (non-object / undefined property/method type runtime errors aside)
 */

$snipsValid = array(
    '$dingus',
    '$dingus->$varProp',
    '$dingus::$staticProp',
    '$dingus->prop',
    '$dingus->prop["foo + bar"]',
    '$dingus->prop[foo + bar]->prop',
    '$dingus->someMethod()->prop',
    'dingus::$staticProp',
    'dingus::someMethod()->prop',
    '${$classVar}->{$varProp}',
    '${$classVar}->$propVar',
);
$snipsInvalid = array(
    'trim($name)',
    '$var+$var',
    '1+1',
    '0',
    '"string"',
    'MYCONST',
    'dingus::MYCONST',
    '$dingus::MYCONST',
    '$dingus->buildObj()',
);

test($snipsValid, true);
test($snipsInvalid, false);

/**
 * This in no way attempts to validate whether snips will parse, logic errors, etc.
 * Nor expected to detected non-object / undefined method/prop type runtime errors
 */
function test($testArray, $expect)
{
    $regExProp = '(\w*|\{\$\w+\})(\[.+?\])*';
    $regEx = '/^
        \s* # ignore leading whitespace
        (
            \$(\w|\{\$\w+\})+                   # var followed by
            (::\$'.$regExProp.'|::\w+\(       # 0 or 1 static prop or method
                .*?                 # method params
            \)(?=::|->))?
            |                       #   OR
            \w+                     # classname followed by
            (::\$'.$regExProp.'|::\w+\(       # static prop or method
                .*?                 # method params
            \)(?=::|->))
        )
        (->'.$regExProp.'(\(       # 0 or more property or methods
            .*?         # method params
        \)(?=::|->))?)*
        (->\$?'.$regExProp.')?
        \s* # ignore trailing whitespace
    $/x';
    $success = true;
    foreach ($testArray as $test) {
        $isVar = preg_match($regEx, $test);
        if ($isVar != $expect) {
            echo 'Test Failed for '.htmlspecialchars($test).' expected '.($expect?'true':'false').'<br />';
            $success = false;
        }
    }
    if ($success) {
        echo 'all is well<br />';
    }
}
```

outputs:

    all is well
    all is well

**Comment by ravage84** [comment]
> This is likely a "is it worth the effort/expense" question

Good point and I tend to say it is not as it is quite an edge case.

A fair code coverage by unit tests will save your ass way better than phpmd could.
